python/langsmith/_internal/_multipart.py

```diff
@@ -0,0 +1,23 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Dict, Iterable, Tuple
+
+MultipartPart = Tuple[str, Tuple[None, bytes, str, Dict[str, str]]]
+
+
+@dataclass
+class MultipartPartsAndContext:
```

Comment:
Do we need all these to be dataclasses? 

I'm sure this is dwarfed by the other improvements, but dataclass creation is much slower than dict/slotted class creation. So while we're on the topic of perf:
```shell
(prompt-optimizer) ➜  langchainplus-sdk git:(main) ✗ python foo.py d # dataclass
Time taken to create 10000000 instances: 6.2248 seconds
(prompt-optimizer) ➜  langchainplus-sdk git:(main) ✗ python foo.py t # typed dict
Time taken to create 10000000 instances: 1.9094 seconds
(prompt-optimizer) ➜  langchainplus-sdk git:(main) ✗ python foo.py slt # regular class with __slots__
Time taken to create 10000000 instances: 1.0796 seconds
(prompt-optimizer) ➜  langchainplus-sdk git:(main) ✗ python foo.py std # stdlib dict
Time taken to create 10000000 instances: 1.0785 seconds
```

Reply:
interesting. relative to the other things I think this might be worth the overhead vs. maintaining init/slot stuff. Could also make these immutable at some point if we want

in the benchmark, we spend 99.8% of the `serialize_run_dict` time on `_dumps_json` and creating these classes is part of the remainder
![ScreenShot 2024-10-31 at 12 55 27PM](https://github.com/user-attachments/assets/5d0afd59-da40-4882-8644-4dc0bb4ad8f5)

Reply:
fair fair